text/0000-prepare-operation.md

```diff
@@ -0,0 +1,385 @@
+- `project.toml` - A project descriptor file specified [here][project-descriptor-spec].
+
+[project-descriptor-spec]: https://github.com/buildpacks/spec/blob/main/extensions/project-descriptor.md
+
+# Motivation
+[motivation]: #motivation
+
+The motiviation for this change was to enable the following goals from being achieved:
+
+  - **Goal 1: Serializing CLI configuration**
+    > As a user, I would like to be able to serialize and share the parameters I use with certain platforms such as `pack`.
+  - **Goal 2: A recognizable file in repositories**
+    > As a user, I would like to be able to recognize, based on the file system, if a project is using Cloud Native Buildpacks.
+  - **Goal 3: Platform recognition of project.toml**
+    > As a user, I would like to ensure that my configuration in `project.toml` is being used.
+
+NOTE: This RFC directly solves for Goal 3, while enabling the solutions for Goals 1 and 2 via a supplimental RFC ([Support for pack.toml][rfc-pr-189]).
+
+[rfc-pr-189]: https://github.com/buildpacks/rfcs/pull/189
+
+# What it is
+[what-it-is]: #what-it-is
+
+The proposal is composed of following changes:
+
+1. [Moving `io.buildpacks` properties to `io.buildpacks.defaults`.](#Namespace-iobuildpacksdefaults)
+2. [A replaceable new phase `prepare` that applies platform configuration.](#Preparer)
+3. [Supporting Cloud Native Buildpack utilities.](#Cloud-Native-Buildpacks-utilities)
+
+## Changes to `io.buildpacks` namespace
+
+### Namespace `io.buildpacks.defaults`
+
+This namespace declares default values that platforms should acknowledge. A majority of properties should be spec'd but we would allow for unspec'd additional properties.
+
+The criteria for spec'd properties in this namespace would be based on whether they affect inputs to the lifecycle phases. For example, the order and environment variables are inputs to the lifecycle. The `builder`, in contrast, is not an input to the lifecycle so it wouldn't meet the criteria.
+
+Therefore the following changes have been made from the [latest schema][pd-02]:
+
+ - Remove `builder`
+
+[pd-02]: https://github.com/buildpacks/spec/blob/extensions/project-descriptor/0.2/extensions/project-descriptor.md#iobuildpacks-optional
+
+### Version `io.buildpacks` namespace
+
+The `io.buildpacks` namespace schema should be able to change without changing the Project Descriptor spec. For this reason, the namespace should define a `schema-version`.
+
+### Schema
+
+See [`io.buildpacks` Namespace Schema](#io-buildpacks-Namespace-Schema) section below.
+
+#### Usages Examples
+
+###### Multi-Platform (pack + kpack)
+
+In the following example we show how various platform namespaces can have their own configuration but properties can be promoted to `io.buildpacks.defaults` if the user's intent is to apply it to all platforms.
+
+```toml
+[io.buildpacks]
+schema-version = "0.2"
+
+###
+# common buildpacks config
+##
+
+[io.buildpacks.defaults]
```

Comment:
**"Naming" Vote: Should we use or not use a `defaults` table?**

It's been discussed that `defaults` in `io.buildpacks.defaults` is unnecessary and that the extra characters aren't worth the effort in trying to convey that these are _only_ defaults and it's really up to the platform to determine which properties get applied.

Please vote for your opinion using the following reactions:

👍 → keep `defaults`
👎 → drop `defaults`

Reply:
RFC has been updated to reflect consensus.